digiroad2-api-oth/src/main/scala/fi/liikennevirasto/digiroad2/MunicipalityApi.scala

```diff
@@ -1,1132 +1,390 @@
-      case RoadWidth.typeId => roadWidthService
-      case MassTransitLane.typeId => massTransitLaneService
-      case NumberOfLanes.typeId => numberOfLanesService
-      case _ => linearAssetService
-    }
-  }
-
-  private def verifyPointServiceToUse(typeId: Int): PointAssetOperations = {
-    typeId match {
-      case Obstacles.typeId  => obstacleService
-      case PedestrianCrossings.typeId   => pedestrianCrossingService
-      case RailwayCrossings.typeId  => railwayCrossingService
-      case TrafficLights.typeId  => trafficLightService
-    }
-  }
-
-  private def getValue(typeId: Int, x : NewAssetValues) : Value = {
-    if (isDynamic(typeId)) {
-      val propertiesDB = assetPropertyService.getProperties(typeId).head
-      DynamicValue(DynamicAssetValue(Seq(DynamicProperty(propertiesDB.publicId, propertiesDB.propertyType, propertiesDB.required, Seq(DynamicPropertyValue(x.properties.map(_.value).head))))))
-    } else
-      NumericValue(x.properties.map(_.value).head.toInt)
-  }
-
-  private def extractNewLinearAssets(typeId: Int, value: JValue) = {
-    AssetTypeInfo.apply(typeId).geometryType match {
-      case "linear" => value.extractOpt[Seq[NewAssetValues]].getOrElse(Nil).map(x => NewLinearAsset(x.linkId, x.startMeasure, x.endMeasure.getOrElse(0), getValue(typeId, x), x.sideCode.getOrElse(SideCode.BothDirections.value), x.geometryTimestamp.getOrElse(vvhClient.roadLinkData.createVVHTimeStamp()), None))
-      case _ => Seq.empty[NewLinearAsset]
-    }
-  }
-
-  private def extractLinearAsset(typeId: Int, value: JValue) = {
-    value.extractOpt[NewAssetValues].map { v =>
-      NewLinearAsset(v.linkId, v.startMeasure, v.endMeasure.getOrElse(0), getValue(typeId, v), v.sideCode.getOrElse(SideCode.BothDirections.value), v.geometryTimestamp.getOrElse(vvhClient.roadLinkData.createVVHTimeStamp()), None)
-    }.get
-  }
-
-  def expireLinearAsset(assetTypeId: Int, assetId: Long, username: String, expired: Boolean): String = {
-    if (linearAssetService.getPersistedAssetsByIds(assetTypeId, Set(assetId.toLong)).isEmpty)
-      halt(NotFound("Asset not found."))
+  def withDynTransaction[T](f: => T): T = OracleDatabase.withDynTransaction(f)
+  def withDynSession[T](f: => T): T = OracleDatabase.withDynSession(f)
+  def awsDao: AwsDao = new AwsDao
 
-    linearAssetService.expireAsset(assetTypeId, assetId, user.username, expired = true).getOrElse("").toString
-  }
+  final val AwsUser = "AwsUpdater"
 
-  def expirePointAsset(assetTypeId: Int, assetId: Long, username: String): Long = {
-    getPointAssetById(assetTypeId, assetId)
-
-    val service = verifyPointServiceToUse(assetTypeId)
-    service.expire(assetId, user.username) match {
-      case 1 => assetId
-      case _ => halt(InternalServerError("Asset not deleted"))
+  private def linkIdValidation(datasetId: String, featureId: Long, linkIds: Set[Long]): (Int, Int) = {
+    val roadLinks = roadLinkService.getRoadsLinksFromVVH(linkIds, false)
```

Comment:
This request to VVH applies two times during the process, one is here, another in the update asset method.
Let's perform this request to VVH before all of this, and the only send the road link needed to inside the methods

Reply:
Applied this suggestion so it call one time per dataset per function(validate and update)